controller/bindings/curl/curl.ml

```diff
@@ -0,0 +1,97 @@
+type error =
+  | UnsuccessfulStatus of int * string
+  | UnreadableStatus of string
+  | ProcessExit of int * string
+  | ProcessKill of int
+  | ProcessStop of int
+
+let pretty_print_error error =
+  match error with
+  | UnsuccessfulStatus (code, body) ->
+      Printf.sprintf "unsuccessful status %d: %s" code body
+
+  | UnreadableStatus body ->
+      Printf.sprintf "unreadable status code %s" body
+
+  | ProcessExit (_, err) ->
+      String.trim err
+
+  | ProcessKill n ->
+      Printf.sprintf "curl killed by signal %d" n
+
+  | ProcessStop n ->
+      Printf.sprintf "curl stopped by signal %d" n
+
+type result =
+  | RequestSuccess of int * string
+  | RequestFailure of error
+
+let exec cmd =
+  let stdout_r, stdout_w = Unix.pipe () in
+  let stderr_r, stderr_w = Unix.pipe () in
+  let%lwt result =
+    Lwt_process.exec
+      ~stdout:(`FD_move stdout_w)
+      ~stderr:(`FD_move stderr_w)
+      cmd
+  in
+  let stdout_input = Lwt_io.of_unix_fd ~mode:Lwt_io.input stdout_r in
+  let stderr_input = Lwt_io.of_unix_fd ~mode:Lwt_io.input stderr_r in
+  let%lwt stdout = Lwt_io.read stdout_input in
+  let%lwt stderr = Lwt_io.read stderr_input in
+  Lwt.return (result, stdout, stderr)
+
+let safe_int_of_string str =
+  try
+    Some (int_of_string str)
+  with
+    Failure _ -> None
+
+let http_code_marker = '|'
+
+let parse_status_code_and_body str =
+  let open Base.Option in
```

Comment:
The standard library miss some usefull functions, which are covered by `base` from jane street:
https://ocaml.janestreet.com/ocaml-core/latest/doc/base/index.html

Discussion about using `base` or `batteries`:
https://www.reddit.com/r/ocaml/comments/azuqpr/base_vs_batteries/